OIPA/iati/parser/tests/test_2_03_parser.py

```diff
@@ -2095,3 +2095,134 @@ def test_activity_result_indicator_document_link_document_category(self):
+        result_reference_XML_element = E(
+            'document-link',
+            **result_reference_attr
+        )
+        try:
+            self.parser_203 \
+            .iati_activities__iati_activity__result__reference(  # NOQA: E501
+            result_reference_XML_element
+            )
+        except RequiredFieldError as inst:
+            self.assertEqual(inst.field, 'vocabulary')
+            self.assertEqual(inst.message,
+                             'required attribute missing')
+
+        # case 2: where 'vocabulary' cannot be found because
+        # dummy 'ResultVocabulary object was not created.
+
+        result_reference_attr = {
+            "vocabulary": '100',
+            "code": '01',
+            "vocabulary-uri": 'www.example.com'
+
+        }
+        result_reference_XML_element = E(
+            'document-link',
+            **result_reference_attr
+        )
+        try:
+            self.parser_203 \
+                .iati_activities__iati_activity__result__reference(
+                    result_reference_XML_element
+                )
+        except FieldValidationError as inst:
+            self.assertEqual(inst.field, 'vocabulary')
+            self.assertEqual(inst.message,
+                             'not found on the accompanying code list')
+
+        # case 3: where 'code' is missing
+
+        result_reference_attr = {
+            "vocabulary": self.result_vocabulary.code,
+            # "code": '01',
+            "vocabulary-uri": 'www.example.com'
+
+        }
+        result_reference_XML_element = E(
+            'document-link',
+            **result_reference_attr
+        )
+        try:
+            self.parser_203 \
+                .iati_activities__iati_activity__result__reference(
+                     result_reference_XML_element
+                )
+        except RequiredFieldError as inst:
+            self.assertEqual(inst.field, 'code')
+            self.assertEqual(inst.message,
+                             'Unspecified or invalid.')
+
+        # case 4: where 'vocabulary_uri' is missing. This case is not tested
+        # because 'vocabulary_uri' is optional.
+
+        # case 5: all is good
+        result_reference_attr = {
+            "vocabulary": self.result_vocabulary.code,
+            "code": '01',
+            "vocabulary-uri": 'www.example.com'
+
+        }
+        result_reference_XML_element = E(
+            'document-link',
+            **result_reference_attr
+        )
+        self.parser_203 \
+            .iati_activities__iati_activity__result__reference(  # NOQA: E501
+                result_reference_XML_element
+            )
+
+        # get 'ResultReference to check if its attributes are correctly stored
+        result_reference = self.parser_203.get_model('ResultReference')
+
+        # check everything is correctly stored
+        self.assertEqual(self.result, result_reference.result)
+        self.assertEqual('01', result_reference.code)
```

Comment:
Also just a small style note, but you could also write `self.assertEqual(result_reference.code, result_reference_attr['code']` or `self.assertEqual(result_reference.code, result_reference_attr.get('code')`